From a4474015eba8b52bfcdb9be9eed6581649029652 Mon Sep 17 00:00:00 2001
From: Marek Vasut <marek.vasut+renesas@mailbox.org>
Date: Sun, 26 Oct 2025 20:37:10 +0100
Subject: [PATCH 2/2] fix(rcar4): assure SCIF and HSCIF clock are always
 enabled

Prior stages before TFA BL31 may shut down the SCIF or HSCIF clock.
Enable the SCIF or HSCIF clock again during console initialization,
to make sure those clock are running and prints can be produced on
serial console if necessary.

Signed-off-by: Marek Vasut <marek.vasut+renesas@mailbox.org>
Change-Id: I9461e72888a939c0bbe5795087a76f1a71c7151d
---
 drivers/renesas/rcar_gen4/scif/scif.c | 27 ++++++++++++++++++++++++++-
 1 file changed, 26 insertions(+), 1 deletion(-)

diff --git a/drivers/renesas/rcar_gen4/scif/scif.c b/drivers/renesas/rcar_gen4/scif/scif.c
index f8eb89df7..786b2d7ea 100644
--- a/drivers/renesas/rcar_gen4/scif/scif.c
+++ b/drivers/renesas/rcar_gen4/scif/scif.c
@@ -14,6 +14,18 @@
 
 #include "rcar_def.h"
 
+/* CPG */
+#define CPG_BASE		0xE6150000UL
+#define CPG_CPGWPR		(CPG_BASE + 0x0000UL)
+#define CPG_CPGWPCR		(CPG_BASE + 0x0004UL)
+#define CPG_MSTPCR5		(CPG_BASE + 0x2D14UL)
+#define CPG_MSTPSR5		(CPG_BASE + 0x2E14UL)
+#define CPG_MSTPSR5_HSCIF0	BIT(14)
+#define CPG_MSTPCR7		(CPG_BASE + 0x2D1CUL)
+#define CPG_MSTPSR7		(CPG_BASE + 0x2E1CUL)
+#define CPG_MSTPSR7_SCIF0	BIT(2)
+#define CPG_MSTPSR7_SCIF3	BIT(4)
+
 /* RST */
 #define RST_BASE		(0xE6160000UL + (RCAR_DOMAIN * 0x4000UL))
 #define RST_MODEMR0		RST_BASE
@@ -29,8 +41,10 @@
 /* SCIF */
 #if (RCAR_LSI == RCAR_S4) /* S4 */
 #define SCIF_BASE	SCIF3_BASE
+#define CPG_MSTPSR7_BIT	CPG_MSTPSR7_SCIF3
 #else
 #define SCIF_BASE	SCIF0_BASE
+#define CPG_MSTPSR7_BIT	CPG_MSTPSR7_SCIF0
 #endif
 #define SCIF_SCFTDR	(SCIF_BASE + 0x000CU)	/*  8 Transmit FIFO data register */
 #define SCIF_SCFSR	(SCIF_BASE + 0x0010U)	/* 16 Serial status register */
@@ -49,7 +63,7 @@
 int console_rcar_init(uintptr_t base_addr, uint32_t uart_clk,
 		      uint32_t baud_rate)
 {
-	uint32_t modemr;
+	uint32_t modemr, mstpcr, mstpsr, mstpbit;
 
 	modemr = ((mmio_read_32(RST_MODEMR0) & RST_MODEMR0_MD31) >> 31U) |
 		 ((mmio_read_32(RST_MODEMR1) & RST_MODEMR1_MD32) << 1U);
@@ -57,10 +71,21 @@ int console_rcar_init(uintptr_t base_addr, uint32_t uart_clk,
 	if (modemr == MODEMR_HSCIF_DLMODE_3000000 ||
 	    modemr == MODEMR_HSCIF_DLMODE_1843200 ||
 	    modemr == MODEMR_HSCIF_DLMODE_921600) {
+		mstpcr = CPG_MSTPCR5;
+		mstpsr = CPG_MSTPSR5;
+		mstpbit = CPG_MSTPSR5_HSCIF0;
 		scif_console_set_regs(HSCIF_HSFSR, HSCIF_HSFTDR);
 	} else {
+		mstpcr = CPG_MSTPCR7;
+		mstpsr = CPG_MSTPSR7;
+		mstpbit = CPG_MSTPSR7_BIT;
 		scif_console_set_regs(SCIF_SCFSR, SCIF_SCFTDR);
 	}
 
+	/* Turn SCIF/HSCIF clock ON. */
+	mmio_clrbits_32(mstpcr, mstpbit);
+	while (mmio_read_32(mstpsr) & mstpbit)
+		;
+
 	return 1;
 }
-- 
2.34.1

